source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.MessageHub/Models/AvailableChargeLinksReceiptData/ChargeLinksCimValidationErrorTextFactory.cs

```diff
@@ -0,0 +1,124 @@
+        private readonly ILogger _logger;
+
+        public ChargeLinksCimValidationErrorTextFactory(
+            ICimValidationErrorTextProvider cimValidationErrorTextProvider,
+            ILoggerFactory loggerFactory)
+        {
+            _cimValidationErrorTextProvider = cimValidationErrorTextProvider;
+            _logger = loggerFactory.CreateLogger(nameof(ChargeLinksCimValidationErrorTextFactory));
+        }
+
+        public string Create(ValidationError validationError, ChargeLinksCommand command)
+        {
+            return GetMergedErrorMessage(validationError, command);
+        }
+
+        private string GetMergedErrorMessage(ValidationError validationError, ChargeLinksCommand chargeLinksCommand)
+        {
+            var errorTextTemplate = _cimValidationErrorTextProvider
+                .GetCimValidationErrorText(validationError.ValidationRuleIdentifier);
+
+            return MergeErrorText(errorTextTemplate, chargeLinksCommand, validationError.TriggeredBy);
+        }
+
+        private string MergeErrorText(
+            string errorTextTemplate,
+            ChargeLinksCommand chargeLinksCommand,
+            string? triggeredBy)
+        {
+            var tokens = CimValidationErrorTextTokenMatcher.GetTokens(errorTextTemplate);
+
+            var mergedErrorText = errorTextTemplate;
+
+            foreach (var token in tokens)
+            {
+                var data = GetDataForToken(token, chargeLinksCommand, triggeredBy);
+                mergedErrorText = mergedErrorText.Replace("{{" + token + "}}", data);
+            }
+
+            return mergedErrorText;
+        }
+
+        private string GetDataForToken(
+            CimValidationErrorTextToken token,
+            ChargeLinksCommand chargeLinksCommand,
+            string? triggeredBy)
+        {
+            // Please keep sorted by CimValidationErrorTextToken
+            return token switch
+            {
+                CimValidationErrorTextToken.ChargeLinkStartDate =>
+                    GetChargeLinkStartDate(chargeLinksCommand, triggeredBy),
+                CimValidationErrorTextToken.DocumentSenderProvidedChargeId =>
+                    GetDocumentSenderProvidedChargeId(chargeLinksCommand, triggeredBy),
+                CimValidationErrorTextToken.MeteringPointId =>
+                    chargeLinksCommand.MeteringPointId,
+                _ => string.Empty,
+            };
+        }
+
+        private string GetChargeLinkStartDate(ChargeLinksCommand chargeLinksCommand, string? triggeredBy)
+        {
+            try
+            {
+                return chargeLinksCommand.ChargeLinks
+                    .Single(p => p.SenderProvidedChargeId == triggeredBy)
+                    .StartDateTime.ToString();
+            }
+            catch (Exception e)
+            {
+                return LogErrorAndReturnErrorMessage(triggeredBy, nameof(ChargeLinkDto.StartDateTime), e);
```

Comment:
Looks like a method with two completely different responsibilities. How about?

```suggestion
                LogError(triggeredBy, nameof(ChargeLinkDto.StartDateTime), e);
                return CimValidationErrorTextTemplateMessages.Unknown;
```